internal/agent/agent.go

```diff
@@ -76,9 +77,44 @@ func (a *Agent) Run(ctx context.Context) error {
 		a.config.ManagementService.Config.AuthInfo.ClientCertificate = filepath.Join(a.config.DataDir, agent_config.DefaultCertsDirName, agent_config.GeneratedCertFile)
 		a.config.ManagementService.Config.AuthInfo.ClientKey = filepath.Join(a.config.DataDir, agent_config.DefaultCertsDirName, agent_config.KeyFile)
 	}
-	publicKey, privateKey, _, err := fcrypto.EnsureKey(deviceReadWriter.PathFor(a.config.ManagementService.AuthInfo.ClientKey))
-	if err != nil {
-		return err
+
+	experimentalFeatures := experimental.NewFeatures()
+	var publicKey crypto.PublicKey
+	var privateKey crypto.PrivateKey
+	var signer crypto.Signer
+	var tpmClient *lifecycle.TpmClient
+
+	// create tpm client and tpm-based keys only if experimental features are enabled
+	if experimentalFeatures.IsEnabled() {
+		a.log.Infof("experimental features enabled: creating TPM client")
+		tpmClient, err := lifecycle.NewTpmClient(a.log)
+		if err != nil {
+			a.log.Infof("tpm is not available: %v", err)
+			tpmClient = nil
+		} else {
+
+			publicKey = tpmClient.GetLocalAttestationPubKey()
+			signer, err = tpmClient.GetSigner()
+			if err != nil {
+				a.log.Errorf("unable to get TPM-based signer for CSR: switching to software keys instead")
+				tpmClient = nil
+			}
```

Comment:
can encapsulate this logic into something where the output is the key and signer etc and the implementation including experimental is abstracted away.